### ISSUE TITLE
API endpoints

### ISSUE THREAD

**Issue by Deraen** [ato/clojars-web#325]
Cljsjs could use richer API to pull out the list of all packages in the group.

/:group-id.json
- List all projects
- Latest version, homepage, clojars url, description
- Download counts?

/:group-id/:artifact-id.json
- Latest version, homepage, clojars url, description
- List of the recent versions
- Download counts?

This is a duplicate to #169 but I thought it best to create a new issue

**Comment by dotemacs** [comment]
We can do this, no problem. 

It's just to bear in mind that this change was done:  #315  for [shields.io](https://shields.io), with this PR https://github.com/badges/shields/pull/427.

I already thought of adding the download counts, as that is already being displayed on the dashboard view for a given clojar.

Before starting on this, can we agree here on the desired output, as per your suggestion above. Something like this for **/:group-id.json/:atifact-id.json**:

```json
{
    "version": "x.y.z",
    "homepage": "http:/...",
    "downloads": "123"
}
```

I'm not sure about 

* the **list of recent versions**. Surely you'd get that if you just followed the URL to the homepage. 

* **clojars-url**, surely that is already 'known' as you'd be accessing **http://clojars.org/some-clojar-name/something.json** as per the #315 we've got **latest-version.json**, which could be renamed to something more meaningful.

As for **/:group-id.json**, the same points as listed above for **list of recent versions** & **clojars-url**, otherwise:

```json
{
    "version": "x.y.z",
    "homepage": "http:/...",
    "artifacts": "list of artifacts",
}
```

and then downloads per artifact... ?

What do you think?

**Comment by Deraen** [comment]
Well, I did already implement this, lets continue talking about this on the PR.

**Comment by tobias** [comment]
Leaving this open until we document it at https://github.com/ato/clojars-web/wiki/Data

**Comment by dotemacs** [comment]
Looking at my work: https://clojars.org/api/groups/pdfboxing how do I populate **scm** & **authors** fields, as they are currently blank?

**Comment by Deraen** [comment]
Interesting, looks like there are problems with custom jars by groupname query.

Another problem I noticed is that is is returning duplicate results in some cases.

E.g. https://clojars.org/api/groups/cljsjs and hashids jar.

**Comment by dotemacs** [comment]
Started with the documentation: https://github.com/ato/clojars-web/wiki/Data#api

Feel free to correct it, add more.

**Comment by Deraen** [comment]
To fix the problems I need proper test data but I'm having problems setting up the db using the new instructions.

If I run `lein migrate` first `clojars.dev.setup` fails with:
```
Exception in thread "main" java.io.FileNotFoundException: data/events/user.clj (No such file or directory), compiling:(/tmp/form-init7032875274853152729.clj:1:72)
```
If I remove data directory and run dev setup, I get:
```
Exception in thread "main" java.sql.SQLException: [SQLITE_ERROR] SQL error or missing database (no such table: jars), compiling:(/tmp/form-init156348020815109138.clj:1:71)
```

**Comment by Deraen** [comment]
Okay managed to setup the db. `clojars.dev.setup` worked the second time.

**Comment by tobias** [comment]
@dotemacs - `:scm` and `:authors` have columns in the db, but are blank. It looks like they are now read from the pom. I'd prefer to not load and parse poms for each api call, at least not without some caching in place. My preference is to drop `:scm` and `:authors` from the response for now.

@Deraen - The duplicates may be caused by each deploy having a row in the db, so an artifact will appear n times, once for each version deployed.

@dotemacs - Thanks for starting the docs. I can add some more detail there.

@Deraen - Sorry about that, the events dir needs to exist, and isn't created automatically or mentioned in the instructions. I'll fix that.

**Comment by tobias** [comment]
I noticed that the api doesn't return a 404 for a missing resource. I'll fix that.

**Comment by Deraen** [comment]
I think I managed to fix duplicates. I added some tests for that also.

https://github.com/ato/clojars-web/pull/335

**Comment by Deraen** [comment]
Duplicates were caused by bad join condition which didn't care about group_name so if there were artifacts in multiple groups with same jar_name the join could result in duplicates.

There might have also been problems in cases where jar only had SNAPSHOT but I think that should also work now.

**Comment by tobias** [comment]
Great, thanks! I just pushed the 404 fix, so I'll pull your changes and rebase on to master. I'll also drop :scm and :authors from the response.

**Comment by tobias** [comment]
@Deraen - I just merged that PR to master. It looks like the query isn't quite right though - it isn't returning the latest version when there are multiple non-snapshot versions of the same jar. There is a failing test in the api integ tests, and I added a new failing test in the api unit tests that demonstrates the issue.

**Comment by Deraen** [comment]
Oh, sorry, I should have noticed the failing integration test.

I opened yet another PR.

**Comment by Deraen** [comment]
It works! https://github.com/cljsjs/cljsjs.github.io/commit/a4429ac34e0e9681d8700b673972d2a5d9737e29

**Comment by tobias** [comment]
Sweet, I guess you've already noticed that I pushed it to production :)

**Comment by tobias** [comment]
I want to make one more change - the /groups data has `:latest_version` and `:latest_release` for each artifact, but the /artifacts data just has `:version`. I'm going to make those consistent (so /artifacts will now include `:latest_*` instead of `:version`) unless there are any objections.

**Comment by Deraen** [comment]
Sounds good.

**Comment by tobias** [comment]
That change is now live.

**Comment by Deraen** [comment]
I updated the wiki with table of formats and their content-types and list of all endpoints with examples.

**Comment by Deraen** [comment]
E.g. https://clojars.org/api/groups/cljsjs seems to load for abour 2-3 seconds. Any ideas if this is caused by Sqlite or what? And is it a problem (for the server)?

**Comment by tobias** [comment]
I noticed last night that it is pretty slow, even for a non-existent group (https://clojars.org/api/groups/asdfasdf takes 1.2s to return a 404). I played with it locally, and I think it's related to the size of the jar table. With a small test db, it's fast (4ms), but with a copy of the production db, the same request takes ~800ms. I tried adding an index on jars just for group_name, but that didn't seem to help.

It's been a long time since I had to optimize sql, so I don't have a feel for what's happening under the hood for the jar query. I wonder if it would be faster to break that in to multiple queries, and combine the results in Clojure. My gut tells me no, since that's what sqlite is designed to do, but I believe we are using the full-java sqlite implementation, which may not be as efficient as the C-based implementation.

I'd like to see if we can improve the speed before we announce it to the world. It hasn't been a problem for the server so far.

**Comment by Deraen** [comment]
https://bitbucket.org/xerial/sqlite-jdbc should use native sqlite.

**Comment by Deraen** [comment]
Instead of 3 joins it would be possible to do 1 join and find the latest release using sub query but in general using joins should be faster.

**Comment by cddr** [comment]
Without spending more than 5 seconds looking at the query, I suspect what's
slowing it down is the "like '%-SNAPSHOT'" because I think that is
unindexable.

**Comment by Deraen** [comment]
I checked the query using explain.

```
1|0|0|SCAN TABLE jars USING COVERING INDEX jars_idx0 (~1000000 rows)
2|0|0|SCAN TABLE jars USING INDEX jars_idx0 (~500000 rows)
3|0|0|SCAN TABLE jars (~1000000 rows)
0|0|1|SCAN SUBQUERY 1 AS l (~100 rows)
0|1|0|SEARCH TABLE jars AS j USING INDEX jars_idx0 (group_name=? AND jar_name=? AND created=?) (~2 rows)
0|2|2|SEARCH SUBQUERY 2 AS r USING AUTOMATIC COVERING INDEX (group_name=? AND jar_name=?) (~3 rows)
0|3|3|SEARCH SUBQUERY 3 AS r2 USING AUTOMATIC COVERING INDEX (created=? AND group_name=? AND jar_name=?) (~7 rows)
0|0|0|USE TEMP B-TREE FOR ORDER BY
```

By adding group_name = ? condition to all join queries, it can be improved if I understand this correctly:
```
1|0|0|SEARCH TABLE jars USING COVERING INDEX jars_idx0 (group_name=?) (~10 rows)
2|0|0|SEARCH TABLE jars USING INDEX jars_idx0 (group_name=?) (~5 rows)
3|0|0|SEARCH TABLE jars USING INDEX jars_idx0 (group_name=?) (~10 rows)
0|0|1|SCAN SUBQUERY 1 AS l (~100 rows)
0|1|0|SEARCH TABLE jars AS j USING INDEX jars_idx0 (group_name=? AND jar_name=? AND created=?) (~2 rows)
0|2|2|SEARCH SUBQUERY 2 AS r USING AUTOMATIC COVERING INDEX (group_name=? AND jar_name=?) (~3 rows)
0|3|3|SEARCH SUBQUERY 3 AS r2 USING AUTOMATIC COVERING INDEX (created=? AND group_name=? AND jar_name=?) (~2 rows)
0|0|0|USE TEMP B-TREE FOR ORDER BY
```

**Comment by Deraen** [comment]
https://github.com/ato/clojars-web/pull/338

**Comment by tobias** [comment]
@Deraen - thanks for the fix, that did the trick. I modified it slightly to also include `jar_name` in the where clause (https://github.com/ato/clojars-web/commit/8064a1c91d2339eda9e51db1d4a6cc086e1b8267), and deployed it to production.

@cddr - thanks for taking a look, that seems obvious once you point it out :)

I'll announce this on clojars-maintainers@. @Deraen - would you like to announce it on clojure@ since this is your baby? If not, I would be happy to do it.

**Comment by Deraen** [comment]
Thanks, I posted the announcement clojars group and made a tweet.
